### PR TITLE
Added 'reparent_child' Function to Node Class.

### PR DESCRIPTION
**Pull request by Eoin-ONeill-Yokai**
This is a proposal for an `reparent_child` method addition to the Node class. This function provides a shorthand method for adding a child that may already have a parent node. This has helped me with reducing code repetition in cases where I want to move a child quickly to another node. An example of what would be done before this patch:

```
func set_up_player_camera_prepatch(player):
	var socket = player.get_node("./CameraSocket")
	var p = camera.get_parent()
	p.remove_child(camera)
	socket.add_child(camera)
	camera.translation = Vector3.ZERO
	camera.rotation = Quat.IDENTITY.get_euler()
```
Becomes the following code after patch:

```
func set_up_player_camera(player):
	var socket = player.get_node("./CameraSocket")
	socket.reparent_child(camera)
	camera.translation = Vector3.ZERO
	camera.rotation = Quat.IDENTITY.get_euler()
```

While the difference may be minor, I believe that it helps create more concise code that allows scripting files to be more concise when passing nodes to other parts of a tree. I believe that the `reparent_child` terminology also makes it clear that this node will become the new parent regardless of the presence of a parent node on a given child. 

## Preservation of Global Transform
For children that have compatible transforms, you can use this method to preserve your global transform using an optional boolean argument.

```
func set_up_player_camera(player):
	var socket = player.get_node("./CameraSocket")
	socket.reparent_child(camera, true)
```

### CODE REVIEW

**Comment by groud** [comment]
I'm am not fully against the idea. But IMHO, this has little use cases and can be workarounded easily.
So I am not sure it is worth merging.

**Comment by akien-mga** [comment]
First of all this I don't think this should be named `adopt_child`, there's no need to further the "family" metaphor and go into the troublesome waters of adoption if we can avoid it. `reparent_child` would be technically more accurate, and thus a better name.

Then I also don't see much use case for this myself, but it's a relatively often requested feature. Yet, if this is implemented it should maybe go further and include an option to preserve the global transform as requested in #21781.

**Comment by groud** [comment]
>  it should maybe go further and include an option to preserve the global transform as requested in #21781.

I think this would be confusing. As global_transform is meaningful only for CanvasItem and Spatial, it would make no sense to add it to the Node class. Also, this this really easy to workaround by registering the global_transform before and applying after the reparenting.

**Comment by akien-mga** [comment]
> Also, this this really easy to workaround by registering the global_transform before and applying after the reparenting.

While that's true, the same applies to the whole reparenting process. If you need to do some bookkeeping for the transform, might as well do the reparenting manually IMO. So either we provide full convenience (which indeed means here that the reparent method would need to be virtual and the transform part handled in CanvasItem and Spatial), or we keep the current status quo.

**Comment by Aaron-Fleisher** [comment]
> "family" metaphor and go into the troublesome waters of adoption if we can avoid i

Are you being serious? This is what you thought of after looking at this PR?

**Comment by groud** [comment]
> Are you being serious? This is what you thought of after looking at this PR?

Please, those kind of interpellation are not constructive feedback. And look a little bit aggressive. :)

Also, I did not mention it but I agree "reparent" is a more adequate term.

**Comment by TheYokai** [comment]
> Then I also don't see much use case for this myself, but it's a relatively often requested feature. Yet, if this is implemented it should maybe go further and include an option to preserve the global transform as requested in #21781.

Yes, this would be relatively easy to add. I will make the suggested changes.

While I agree it's really easy to work around this issue, I basically made this change for my own use because I was sick of the boilerplate when it came to quickly moving nodes to different parts of the tree. I've been using it for a while and I personally enjoy it, I was mostly curious how others felt about it.

I think a preserving global transform option would be an easy addition and would help to reduce even more boiler plate code.

**Comment by reduz** [comment]
I am against this. There is absolutely nothing wrong with writing more code in this case, it's not like you are writing this a hundred times in a project.

**Review by reduz — COMMENTED** [review]
Sorry, I don't approve this. You are not going to need this more than a couple of times in your project. 

It's not even a common use case, so adding a specialized function is unnecessary.

**Comment by TheYokai** [comment]
I have still gone ahead with the proposed changes to allow for preservation of global_transform settings. I've also changed the name of the function and fixed a bug.

I will keep this branch updated with refinements to the design as necessary. Truth be told, I don't mind continuing to cherrypick this change into my own projects for my own use. I simply find it decent utility that I decided might be nice to push upstream when I was asked about it personally.

**Comment by eon-s** [comment]
In my experience, reparenting always look like a bad practice/design but sometimes may be inevitable, if you can provide examples where is hard to do a clean reparenting maybe you will be able to convince other devs.

Complex situations may involve physics, inputs, signals.

ps: in case something like this is added, I'm in favor of the `reparent` term too.

**Comment by Aaron-Fleisher** [comment]
> As for the name, I really don't mind changing the name to something that avoids stepping in any troublesome waters.

Don't even worry about that :), there is nothing wrong with using `adopt_child` in code, and it has nothing to do with adoption centers or families.  `reparent` does make more sense in your case, however (IMO)

I appreciate your PRs, especially the RichTextLabel. You don't deserve this kind of response at all, you did nothing wrong.

**Comment by Naryosha** [comment]
TL;DR I agree with @eon-s. Please convince us with use cases.

<details>

This is my own opinion based on my own experience with Godot. You're welcome to disagree.

I have never really run into a circumstance where I need to `reparent` in runtime. Fortunately, every time I thought I need to reparent, I could find a more straightforward solution that doesn't involve it. Reparenting is hacky and it exposes more undefined behaviors as some node's properties are derived from parent and you have to choose what to keep carefully. IMO reparenting is a hassle in itself and simply adding a method won't solve it.

For tools development, I guess you move nodes around more often, although I think it's not any more complex than `remove_child` and `add_child`. Those methods do more or less what you would expect and you don't have to worry about parent-derived properties.

### About the syntax itself

```gdscript
parent.reparent_child(child, true)
```
It is a victim of [boolean trap](https://ariya.io/2011/08/hall-of-api-shame-boolean-trap) and would add more confusion when you come back to the code after a while without proper commenting, which negates the effort to keep the code concise.

From what I've gathered in the past years, Godot API design was always about verbosity and clarity. We'd rather have longer code than unclear code. (Although you might be able to find counter-examples, especially in GDScript built-in functions, it's often an older code and doesn't represent the design goal really well.)

### About feature creep

If I search for `feature proposal` tag in the closed/unmerged PR, I'll see some of the PR that make me think "Ah. If this was merged it'd be really handy that time when I wanted to do blahblahblah..." but I was doing just fine without it.

Every piece of code we merge become a burden to maintainers, uninterested users, and even players if the binary is not optimized. We have to discuss extensively about whether it'd be useful in general.

### About naming

It's not really the choice of Godot devs. They simply follow the trend. Some may find some terminology more uncomfortable than others, and avoiding them is a good idea.

"Assume positive intent." :)

</details>

**Comment by KoBeWi** [comment]
> There is absolutely nothing wrong with writing more code in this case, it's not like you are writing this a hundred times in a project.

The are LOTS of methods that you don't use hundred of times in a project, and yet, they are still there. If this functionality can be achieved relatively simply, I don't see why it shouldn't be in game. It's more useful than it seems. In almost every of my projects, I write a utility script, which contains, among others, the reparent method. It's really useful e.g. when you attach a level scene, but want the objects to be inside game state (because YSort). Or want to e.g. move items between containers in Node-based inventory. Or want an object to stick to another i.e. become it's child upon collision. There are many examples for use of reparenting.

> I think this would be confusing.

Confusing? You know what is confusing? Local position. Questions like "my node disappeared" are asked quite frequently. People manage to change the parent of a node, but have no idea that they have to keep the old global position. And personally, every time I have to do this it's just awkward, because you need to introduce a temporary variable (reparent's implementation would probably do the same, but whatever). If anything, the current "reparenting" is confusing.

When I created the issue linked by akien-mga, it was not because I randomly thought that since I just wrote it, it should be part of the engine. I created the issue, because I had to implement the method myself numerous times and there are lots of people who don't even know how to do it (let's ignore that they should just learn how to do anything). I bet many people will agree.

**Comment by TheYokai** [comment]
This has received a lot more traffic than I thought it would, so let me start from the top.

@eon-s 
> In my experience, reparenting always look like a bad practice/design

I understand that there are some cases where reparenting may not be the best solution, but it certainly isn't a bad practice in its own right. In fact, I believe that Godot is designed in a way where there's not really a clear 'one size fits all' solution that should be considered a go-to for every project. I believe that it really depends on how you are using node data and what your plans are for your game/app structure. For instance, node reparenting is absolutely a must in cases where you wish to recycle scene instances (instead of instantiating a scene multiple times) and can also be very handy when wanting to make use of local coordinates. Specifically, my current use case uses  'socket' nodes which allow you to attach meshes to one another, or also construct modular structures which are designed to socket into one another. I also use these sockets to quickly move the camera between desired targets (with interpolation being controlled by a camera script.) I also think it is useful in cases where you have 'spawner' entities that are responsible for spawning scenes. Also...

@avencherus 
>It comes up often enough for me. I write a lot of editor tools for my projects, and the code for moving things around would tighten up a bit, and be more readable if there were a reparent() method.

This is exactly how I feel as well. While it may be something that may not happen often in an actual runtime environment, it does come up pretty often when you are trying to make an editor tool that manages node data behind the scenes. 


>It is useful, and not project specific. Given the rest the functions on the Node it seems like it should've already been included from the start.

I agree -- however...

>I would agree it isn't often enough that its breaking my wrist to write the extra lines, but at the same time I don't understand the resistance on the grounds of preventing feature creep? Is that the reasoning? I don't think these are the kinds of small additions that have to be guarded against.

To play devils advocate a bit, I actually understand the logic behind being sceptical toward a change to the api of a low level base class. Simply by adding a `preserve_global_transform` flag, I've already needed to provide custom behaviour in the case that users are using `Spatial` and `Node2D` derived classed. However, this adds to burden of maintenance (these functions will have to be updated whenever a major change happens to transform data) and also adds complexity to all of the classes that derive from these classes. And then there's going to be the inevitable situation where people want more data to be inherited from the parent (CanvasItem bounds, Node Visibility, Whatever Else) that would extend the complexity across (basically) every class.

So the problem isn't so much feature creep as much as it is resistance toward maintaining *another* function that does something similar but with a different implementation. Having said that, I feel that with the proper scope, a simple utility method for removing-and-adding a child would be easily maintainable since most of the actual code surrounding it is simply error handling. 

@Naryosha 
> It is a victim of boolean trap and would add more confusion when you come back to the code after a while without proper commenting, which negates the effort to keep the code concise.

Yes, but frankly that hasn't actually stopped any other part of the code from using booleans as optional flags. So if we're to make a rule about this, it should be consistent across the entire API. 

>Reparenting is hacky and it exposes more undefined behaviors as some node's properties are derived from parent and you have to choose what to keep carefully.

Honestly, if done well, you can avoid these undefined behaviors by using `_notification` to handle these types of use cases. `NOTIFICATION_PARENTED` and `NOTIFICATION_UNPARENTED` are good for these specific scenarios and can help resolve any undefined behaviour. The better argument against node reparenting is that it makes it basically impossible to use `NodePath` to navigate to a desired node if it is moving about. However, that's very specific to the design of the application in question. I would never reparent a node that I was planning to use NodePath to access, with or without this function. 

Lastly...
>You don't deserve that kind of response at all, you did nothing wrong.

I feel the need to point out that people shouldn't really be taking discussions around any of this personally or think of it as such. I believe it's healthy for open source project that PRs are talked about, debated, and have all opinions shared. I'm actually very happy to have a discussion around this, and I actually went into this PR knowing full well that this would be a topic that needs heavy discussion. I simply wanted to share my own private solution and hear some opinions on why this isn't already in the API.

**Comment by PLyczkowski** [comment]
@reduz 
> 
> 
> I am against this. There is absolutely nothing wrong with writing more code in this case, it's not like you are writing this a hundred times in a project.

Maybe I'm doing something not recommended, but I'm reparenting nodes quite often. I'd say it's quite possible that I'll use reparenting a hundred times or more in a big project..

Example use cases for me:
- Reparenting a Panel between Controls for achieving switching GUI elements position between predefined positions - like a popup that goes in different directions.
- Loading a template scene and reparenting nodes from it to the game - like loading an enemy wave scene and reparenting enemies to different positions in the game.
- Tweening Controls between areas that end with reparent - like drawing/discarding cards in a card game.
- Using Nodes as abstractions and reparenting them to indicate their state - like an effect queue, where an Effect extends a Node and is reparented to QueuedEffects Node, and after executing reparented to EffectHistory Node.

Control whether position is preserved or not during reparenting (On Node2D/Control/Spatial) also sounds very useful.

**Comment by TheYokai** [comment]
Would implementing this as a property (as described in #24433) be a better alternative for people? Some of the 'smart' behaviour introduced later in this patch would probably have to be changed or removed, but I may make for an even cleaner to be able to do something like:

```
some_node.parent = new_parent
```

and would also allow for exposure to signals / slots so that you could have your own custom parent-reassign behavior within a given node?

**Comment by Eoin-ONeill-Yokai** [comment]
I'm going to go ahead and close this PR because of the new Godot Proposals ( godotengine/godot-proposals ) repository. Those you feel strongly about this feature (maybe @PLyczkowski @KoWeBi @avencherus and others?) should feel free to open an issue there and link to this PR for previous discussion. 

I'm not sure how the migration from this board to the other is supposed to work, but closing this PR now is probably for the better and can be resubmitted at any time.
